File: ts-backend/src/videoAnalysis.ts
import fs from "node:fs/promises";
import path from "node:path";
import os from "node:os";

import ffmpeg from "fluent-ffmpeg";
import ffmpegPath from "ffmpeg-static";
import Jimp from "jimp";

ffmpeg.setFfmpegPath(ffmpegPath as string);

export type AnalysisResult = {
  report_id: string;
  created_at: string;
  source_filename: string;
  total_frames: number;
  fps: number;
  duration_seconds: number;
  motion_events: number;
  motion_score_mean: number;
  motion_score_std: number;
};

const scoreMotion = async (prevPath: string, nextPath: string) => {
  const prevImage = await Jimp.read(prevPath);
  const nextImage = await Jimp.read(nextPath);
  const diff = Jimp.diff(prevImage, nextImage);
  return diff.percent * 255;
};

const countPeaks = (scores: number[], threshold: number) => {
  let count = 0;
  let armed = true;
  for (const score of scores) {
    if (armed && score >= threshold) {
      count += 1;
      armed = false;
    } else if (!armed && score < threshold * 0.7) {
      armed = true;
    }
  }
  return count;
};

const parseFps = (rate?: string) => {
  if (!rate) return 30;
  const [num, den] = rate.split("/").map(Number);
  if (!den || Number.isNaN(num) || Number.isNaN(den)) return 30;
  return num / den;
};

export const analyzeVideo = async (
  videoPath: string,
  reportPath: string,
): Promise<AnalysisResult> => {
  const metadata = await new Promise<ffmpeg.FfprobeData>((resolve, reject) => {
    ffmpeg.ffprobe(videoPath, (err, data) => {
      if (err) reject(err);
      else resolve(data);
    });
  });

  const videoStream = metadata.streams.find((stream) => stream.codec_type === "video");
  const fps = parseFps(videoStream?.r_frame_rate);
  const durationSeconds = Number(videoStream?.duration || metadata.format.duration || 0);
  const totalFrames = Math.round(durationSeconds * fps);

  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "sports-motion-"));
  const framePattern = path.join(tempDir, "frame-%04d.jpg");
  const sampleFps = 2;

  await new Promise<void>((resolve, reject) => {
    ffmpeg(videoPath)
      .outputOptions([`-vf fps=${sampleFps}`])
      .output(framePattern)
      .on("end", () => resolve())
      .on("error", (error) => reject(error))
      .run();
  });

  const frameFiles = (await fs.readdir(tempDir))
    .filter((name) => name.endsWith(".jpg"))
    .sort()
    .map((name) => path.join(tempDir, name));

  const scores: number[] = [];
  for (let index = 1; index < frameFiles.length; index += 1) {
    const score = await scoreMotion(frameFiles[index - 1], frameFiles[index]);
    scores.push(score);
  }

  await Promise.all(frameFiles.map((file) => fs.unlink(file)));
  await fs.rmdir(tempDir);

  if (!scores.length) scores.push(0);

  const mean = scores.reduce((sum, value) => sum + value, 0) / scores.length;
  const variance =
    scores.reduce((sum, value) => sum + (value - mean) ** 2, 0) / scores.length;
  const std = Math.sqrt(variance);
  const threshold = mean + std * 1.2;

  const result: AnalysisResult = {
    report_id: path.basename(reportPath, ".json"),
    created_at: new Date().toISOString(),
    source_filename: path.basename(videoPath),
    total_frames: totalFrames,
    fps,
    duration_seconds: durationSeconds,
    motion_events: countPeaks(scores, threshold),
    motion_score_mean: mean,
    motion_score_std: std,
  };

  await fs.writeFile(reportPath, JSON.stringify(result, null, 2), "utf-8");
  return result;
};

File: python-backend/app/video_analysis.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
import json
from pathlib import Path
from typing import Any

import cv2
import numpy as np


@dataclass
class AnalysisResult:
    report_id: str
    created_at: str
    source_filename: str
    total_frames: int
    fps: float
    duration_seconds: float
    motion_events: int
    motion_score_mean: float
    motion_score_std: float

    def to_dict(self) -> dict[str, Any]:
        return {
            "report_id": self.report_id,
            "created_at": self.created_at,
            "source_filename": self.source_filename,
            "total_frames": self.total_frames,
            "fps": self.fps,
            "duration_seconds": self.duration_seconds,
            "motion_events": self.motion_events,
            "motion_score_mean": self.motion_score_mean,
            "motion_score_std": self.motion_score_std,
        }


def _score_motion(prev_gray: np.ndarray, gray: np.ndarray) -> float:
    diff = cv2.absdiff(prev_gray, gray)
    return float(np.mean(diff))


def _count_peaks(scores: list[float], threshold: float) -> int:
    count = 0
    armed = True
    for score in scores:
        if armed and score >= threshold:
            count += 1
            armed = False
        elif not armed and score < threshold * 0.7:
            armed = True
    return count


def analyze_video(video_path: Path, report_path: Path) -> AnalysisResult:
    capture = cv2.VideoCapture(str(video_path))
    if not capture.isOpened():
        raise ValueError(f"Unable to open video: {video_path}")

    fps = capture.get(cv2.CAP_PROP_FPS) or 30.0
    total_frames = int(capture.get(cv2.CAP_PROP_FRAME_COUNT) or 0)

    prev_gray = None
    scores: list[float] = []

    while True:
        success, frame = capture.read()
        if not success:
            break
        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        if prev_gray is not None:
            scores.append(_score_motion(prev_gray, gray))
        prev_gray = gray

    capture.release()

    if not scores:
        scores = [0.0]

    score_mean = float(np.mean(scores))
    score_std = float(np.std(scores))
    threshold = score_mean + score_std * 1.2

    motion_events = _count_peaks(scores, threshold)
    duration_seconds = total_frames / fps if fps else 0.0

    result = AnalysisResult(
        report_id=report_path.stem,
        created_at=datetime.utcnow().isoformat() + "Z",
        source_filename=video_path.name,
        total_frames=total_frames,
        fps=float(fps),
        duration_seconds=float(duration_seconds),
        motion_events=motion_events,
        motion_score_mean=score_mean,
        motion_score_std=score_std,
    )

    report_path.write_text(
        json.dumps(result.to_dict(), ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    return result
